File: src/adapters/zvt_adapter.js
/**
 * ZVT Adapter - Parses ZVT (ZahlVerkehrsTerminal) messages
 * 
 * ZVT is the protocol used in German payment terminals.
 * 
 * Message Structure:
 * - CTRL (2 bytes): Message type/command code
 * - Length (1-3 bytes): Payload length
 *   - 0x00-0xFE: Direct length value
 *   - 0xFF: Extended length (next 2 bytes)
 * - BMP Fields: Tag-Length-Value encoded data
 *   - Each BMP field: 1-byte tag, 1-byte length, variable value
 * 
 * EMV TLV data is embedded in BMP tags:
 * - 0x9A: Transaction-related EMV data
 * - 0xAA: Additional EMV data
 * - 0x3B: Card-related EMV data
 */

const TLVParser = require('../core/tlv_parser');

// ZVT CTRL codes
const CTRL_CODES = {
  '0601': 'Authorisation',
  '060F': 'Authorisation Response',
  '061E': 'End of Day',
  '060B': 'Status Enquiry',
  '068A': 'Print Line',
  '8000': 'ACK',
  '8400': 'Abort',
};

// BMP tags that contain EMV TLV data
const EMV_BMP_TAGS = [0x9A, 0xAA, 0x3B];

class ZVTAdapter {
  /**
   * Parse a ZVT message buffer
   * 
   * @param {Buffer} buffer - Raw ZVT message bytes
   * @returns {Object} Parsed message with ctrl, length, and bmpFields
   */
  static parse(buffer) {
    let offset = 0;

    // Parse CTRL (2 bytes)
    const ctrl = buffer.slice(0, 2).toString('hex').toUpperCase();
    offset += 2;

    // Parse Length (1-3 bytes)
    let length;
    let lengthBytes;
    
    if (buffer[offset] === 0xFF) {
      // Extended length: 0xFF + 2 bytes
      length = (buffer[offset + 1] << 8) | buffer[offset + 2];
      lengthBytes = 3;
    } else {
      // Direct length
      length = buffer[offset];
      lengthBytes = 1;
    }
    offset += lengthBytes;

    // Parse BMP fields
    const bmpFields = [];
    const payloadEnd = offset + length;

    while (offset < payloadEnd) {
      const bmpTag = buffer[offset];
      const bmpLength = buffer[offset + 1];
      offset += 2;

      const bmpValue = buffer.slice(offset, offset + bmpLength);
      offset += bmpLength;

      bmpFields.push({
        tag: bmpTag.toString(16).toUpperCase().padStart(2, '0'),
        length: bmpLength,
        value: bmpValue,
      });
    }

    return {
      ctrl,
      ctrlName: CTRL_CODES[ctrl] || 'Unknown',
      length,
      bmpFields,
    };
  }

  /**
   * Extract EMV TLV data from BMP fields
   * 
   * BMP tags 0x9A, 0xAA, and 0x3B contain EMV TLV data.
   * 
   * @param {Object} message - Parsed ZVT message
   * @returns {TLVNode[]} Array of parsed TLV trees
   */
  static extractEMVTLV(message) {
    const tlvTrees = [];

    for (const bmp of message.bmpFields) {
      const tagValue = parseInt(bmp.tag, 16);
      
      if (EMV_BMP_TAGS.includes(tagValue)) {
        // Parse the BMP value as EMV TLV
        const nodes = TLVParser.parse(bmp.value);
        tlvTrees.push(...nodes);
      }
    }

    return tlvTrees;
  }

  /**
   * Serialize a ZVT message
   * 
   * @param {Object} message - Message object with ctrl and bmpFields
   * @returns {Buffer} Serialized ZVT message
   */
  static serialize(message) {
    const parts = [];

    // CTRL (2 bytes)
    const ctrlBuffer = Buffer.from(message.ctrl, 'hex');
    parts.push(ctrlBuffer);

    // Calculate payload length
    let payloadLength = 0;
    for (const bmp of message.bmpFields) {
      payloadLength += 2 + bmp.value.length; // Tag(1) + Length(1) + Value
    }

    // Length encoding
    let lengthBuffer;
    if (payloadLength <= 0xFE) {
      lengthBuffer = Buffer.from([payloadLength]);
    } else {
      // Extended length
      lengthBuffer = Buffer.from([
        0xFF,
        (payloadLength >> 8) & 0xFF,
        payloadLength & 0xFF,
      ]);
    }
    parts.push(lengthBuffer);

    // BMP fields
    for (const bmp of message.bmpFields) {
      const tagBuffer = Buffer.from([parseInt(bmp.tag, 16)]);
      const lengthBuffer = Buffer.from([bmp.value.length]);
      parts.push(tagBuffer);
      parts.push(lengthBuffer);
      parts.push(bmp.value);
    }

    return Buffer.concat(parts);
  }

  /**
   * Get CTRL code name
   * 
   * @param {string} ctrl - CTRL code in hex
   * @returns {string} CTRL code name
   */
  static getCtrlName(ctrl) {
    return CTRL_CODES[ctrl] || 'Unknown';
  }

  /**
   * Check if BMP tag contains EMV data
   * 
   * @param {string} bmpTag - BMP tag in hex
   * @returns {boolean} True if tag contains EMV data
   */
  static isEMVBmpTag(bmpTag) {
    const tagValue = parseInt(bmpTag, 16);
    return EMV_BMP_TAGS.includes(tagValue);
  }
}

module.exports = ZVTAdapter;

File: src/emv_tlv/adapters/zvt_adapter.py
"""
ZVT Adapter - Parses ZVT (ZahlVerkehrsTerminal) messages.

ZVT is the protocol used in German payment terminals.

Message Structure:
- CTRL (2 bytes): Message type/command code
- Length (1-3 bytes): Payload length
  - 0x00-0xFE: Direct length value
  - 0xFF: Extended length (next 2 bytes)
- BMP Fields: Tag-Length-Value encoded data
  - Each BMP field: 1-byte tag, 1-byte length, variable value

EMV TLV data is embedded in BMP tags:
- 0x9A: Transaction-related EMV data
- 0xAA: Additional EMV data
- 0x3B: Card-related EMV data
"""

from emv_tlv.core.tlv_parser import TLVParser

# ZVT CTRL codes
_CTRL_CODES: dict[str, str] = {
    "0601": "Authorisation",
    "060F": "Authorisation Response",
    "061E": "End of Day",
    "060B": "Status Enquiry",
    "068A": "Print Line",
    "8000": "ACK",
    "8400": "Abort",
}

# BMP tags that contain EMV TLV data
_EMV_BMP_TAGS = {0x9A, 0xAA, 0x3B}


class ZVTAdapter:
    """Adapter for parsing and serializing ZVT messages."""

    @staticmethod
    def parse(data: bytes) -> dict:
        """
        Parse a ZVT message buffer.

        Args:
            data: Raw ZVT message bytes

        Returns:
            Parsed message dict with ctrl, ctrl_name, length, and bmp_fields
        """
        offset = 0

        # Parse CTRL (2 bytes)
        ctrl = data[0:2].hex().upper()
        offset += 2

        # Parse Length (1-3 bytes)
        if data[offset] == 0xFF:
            # Extended length: 0xFF + 2 bytes
            length = (data[offset + 1] << 8) | data[offset + 2]
            length_bytes = 3
        else:
            # Direct length
            length = data[offset]
            length_bytes = 1
        offset += length_bytes

        # Parse BMP fields
        bmp_fields: list[dict] = []
        payload_end = offset + length

        while offset < payload_end:
            bmp_tag = data[offset]
            bmp_length = data[offset + 1]
            offset += 2

            bmp_value = data[offset : offset + bmp_length]
            offset += bmp_length

            bmp_fields.append({
                "tag": f"{bmp_tag:02X}",
                "length": bmp_length,
                "value": bmp_value,
            })

        return {
            "ctrl": ctrl,
            "ctrl_name": _CTRL_CODES.get(ctrl, "Unknown"),
            "length": length,
            "bmp_fields": bmp_fields,
        }

    @staticmethod
    def extract_emv_tlv(message: dict) -> list:
        """
        Extract EMV TLV data from BMP fields.

        BMP tags 0x9A, 0xAA, and 0x3B contain EMV TLV data.

        Args:
            message: Parsed ZVT message dict

        Returns:
            List of parsed TLVNode trees
        """
        tlv_trees: list = []

        for bmp in message["bmp_fields"]:
            tag_value = int(bmp["tag"], 16)

            if tag_value in _EMV_BMP_TAGS:
                # Parse the BMP value as EMV TLV
                nodes = TLVParser.parse(bmp["value"])
                tlv_trees.extend(nodes)

        return tlv_trees

    @staticmethod
    def serialize(message: dict) -> bytes:
        """
        Serialize a ZVT message.

        Args:
            message: Message dict with ctrl and bmp_fields

        Returns:
            Serialized ZVT message bytes
        """
        parts: list[bytes] = []

        # CTRL (2 bytes)
        ctrl_bytes = bytes.fromhex(message["ctrl"])
        parts.append(ctrl_bytes)

        # Calculate payload length
        payload_length = 0
        for bmp in message["bmp_fields"]:
            payload_length += 2 + len(bmp["value"])  # Tag(1) + Length(1) + Value

        # Length encoding
        if payload_length <= 0xFE:
            length_bytes = bytes([payload_length])
        else:
            length_bytes = bytes([
                0xFF,
                (payload_length >> 8) & 0xFF,
                payload_length & 0xFF,
            ])
        parts.append(length_bytes)

        # BMP fields
        for bmp in message["bmp_fields"]:
            tag_byte = bytes([int(bmp["tag"], 16)])
            len_byte = bytes([len(bmp["value"])])
            parts.append(tag_byte)
            parts.append(len_byte)
            parts.append(bmp["value"])

        return b"".join(parts)

    @staticmethod
    def get_ctrl_name(ctrl: str) -> str:
        """Get CTRL code name."""
        return _CTRL_CODES.get(ctrl, "Unknown")

    @staticmethod
    def is_emv_bmp_tag(bmp_tag: str) -> bool:
        """Check if BMP tag contains EMV data."""
        tag_value = int(bmp_tag, 16)
        return tag_value in _EMV_BMP_TAGS